File: cpp/write_async.cpp
#include <iostream>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <string>
#include <locale>
#include <codecvt>
#include <cerrno>

using namespace std;

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: " << argv[0] << " <port>" << endl;
        return 1;
    }
    string portName = argv[1];
    int serialPort = open(portName.c_str(), O_WRONLY | O_NOCTTY | O_NONBLOCK);
    
    if (serialPort == -1) {
        cerr << "Error opening serial port: " << portName << endl;
        return 1;
    }

    termios tty;
    tcgetattr(serialPort, &tty);

    cfsetispeed(&tty, B115200);
    cfsetospeed(&tty, B115200);

    tty.c_cflag &= ~PARENB;
    tty.c_cflag &= ~CSTOPB;
    tty.c_cflag &= ~CSIZE;
    tty.c_cflag |= CS8;
    tty.c_cflag &= ~CRTSCTS;

    tty.c_cc[VMIN] = 1;
    tty.c_cc[VTIME] = 0;

    tcsetattr(serialPort, TCSANOW, &tty);

    string dataToWrite;
    wstring_convert<codecvt_utf8_utf16<char16_t>, char16_t> converter;
    int bytesWritten;

    while (true) {
        cout << "Enter a message to send: ";
        getline(cin, dataToWrite);

        if (dataToWrite.empty()) continue;

        u16string utf16Data = u"\uFEFF";
        utf16Data += converter.from_bytes(dataToWrite);

        bytesWritten = write(serialPort, utf16Data.c_str(), utf16Data.size() * sizeof(char16_t));

        if (bytesWritten == -1) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                cerr << "Write buffer full, try again later." << endl;
            } else {
                cerr << "Error writing to serial port" << endl;
            }
        }
    }
    
    close(serialPort);
    return 0;
}

File: rust/src/bin/write_async.rs
use cable_rust::serial::*;
use std::io::{self, BufRead, Write, ErrorKind};
use std::fs::OpenOptions;
use std::os::unix::fs::OpenOptionsExt;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <port>", args[0]);
        std::process::exit(1);
    }
    let port_str = &args[1];

    configure_serial(port_str);

    let mut file = match OpenOptions::new()
        .write(true)
        .custom_flags(O_NONBLOCK)
        .open(port_str)
    {
        Ok(f) => f,
        Err(e) => {
            eprintln!("Error opening serial port: {}: {}", port_str, e);
            std::process::exit(1);
        }
    };

    let stdin = io::stdin();
    let mut reader = stdin.lock();

    loop {
        print!("Enter a message to send: ");
        io::stdout().flush().unwrap();

        let mut data_to_write = String::new();
        if reader.read_line(&mut data_to_write).is_err() || data_to_write.is_empty() {
            break;
        }

        let data_to_write = data_to_write.trim_end();
        if data_to_write.is_empty() {
            continue;
        }

        match file.write_all(data_to_write.as_bytes()) {
            Ok(_) => (),
            Err(e) if e.kind() == ErrorKind::WouldBlock => {
                eprintln!("Write buffer full, try again later.");
            }
            Err(e) => {
                eprintln!("Error writing to serial port: {}", e);
            }
        }
    }
}
